### PR TITLE
Slowly introduce ops to be tested by test_numpy_ref on MPS backend

### PR DESCRIPTION
**Pull request by aslvrstn**
Enable a test that would have caught https://github.com/pytorch/pytorch/issues/86239

Prior to the fix for that bug, this test fails with

```
_____________________________ TestCommonMPS.test_numpy_ref_mps_where_mps_float32 _____________________________
Traceback (most recent call last):
  File "/Users/alex/git/pytorch/test/test_ops.py", line 197, in test_numpy_ref_mps
    self.compare_with_reference(
  File "/Users/alex/git/pytorch/torch/testing/_internal/common_utils.py", line 2366, in compare_with_reference
    actual = torch_fn(t_inp, *t_args, **t_kwargs)
  File "/Users/alex/git/pytorch/torch/testing/_internal/opinfo/core.py", line 1068, in __call__
    return self.op(*args, **kwargs)
  File "/Users/alex/git/pytorch/torch/testing/_internal/common_methods_invocations.py", line 15167, in <lambda>
    op=lambda self, condition, other: torch.where(condition, self, other),
RuntimeError: 0'th index 3 of x tensor does not match the other tensors
```

### CODE REVIEW

**Comment by aslvrstn** [comment]
Incidentally I think I need to fix another bug before I can merge. Even after the previous fix, this now fails with

```
_____________________________ TestCommonMPS.test_numpy_ref_mps_where_mps_float32 _____________________________
Traceback (most recent call last):
  File "/Users/alex/git/pytorch/test/test_ops.py", line 197, in test_numpy_ref_mps
    self.compare_with_reference(
  File "/Users/alex/git/pytorch/torch/testing/_internal/common_utils.py", line 2366, in compare_with_reference
    actual = torch_fn(t_inp, *t_args, **t_kwargs)
  File "/Users/alex/git/pytorch/torch/testing/_internal/opinfo/core.py", line 1068, in __call__
    return self.op(*args, **kwargs)
  File "/Users/alex/git/pytorch/torch/testing/_internal/common_methods_invocations.py", line 15167, in <lambda>
    op=lambda self, condition, other: torch.where(condition, self, other),
RuntimeError: expected scalar type long long but found float
```

**Comment by aslvrstn** [comment]
Could someone with permissions please add ciflow/mps to this? Thank you! (Am I correct in deducing that that's gated on permissions?)

**Comment by mruberry** [comment]
> Could someone with permissions please add ciflow/mps to this? Thank you! (Am I correct in deducing that that's gated on permissions?)

Added the ciflow/mps label

File: test/test_ops.py [hunk 1]
```diff
@@ -97,6 +98,22 @@
         op_db,
     )
 )
+# This list specifies the ops that should be skipped by `test_numpy_ref_mps`. Currently a large chunk of
+# `op_db`'s tests fail, but as MPS is brought in line with numpy, this list should disappear.
+_ops_failing_on_mps = [
```
Comment (mruberry) [inline test/test_ops.py:103]: OpInfos have a `skips` attribute and a decorator to skip the test can be added there. Adding the skips to the OpInfos keeps all the test data related to a particular operator in one place.

File: test/test_ops.py [hunk 2]
```diff
@@ -172,6 +189,33 @@ def test_numpy_ref(self, device, dtype, op):
...
             torch.set_default_dtype(cur_default)
 
+    # Same as above `test_numpy_ref`, but with different settings required for the MPS backend
+    # When MPS becomes more consistent, this can probably be merged with the above using
+    # `@dtypesIfMPS(torch.float32)`, but for now, the assertions themselves need to be loosened
+    @unittest.skipIf(TEST_WITH_ASAN, "Skipped under ASAN")
+    @onlyMPS
+    @suppress_warnings
+    # MPS only supports float32
+    @ops(_ref_test_ops, allowed_dtypes=(torch.float32,))
```
Comment (mruberry) [inline test/test_ops.py:199]: Should this be `float64`? In the test the default dtype is set to `double` to be more consistent with NumPy, but the test is currently run in `float32`.

File: torch/testing/_internal/common_device_type.py [hunk 3]
```diff
@@ -609,9 +612,13 @@ def instantiate_device_type_tests(generic_test_class, scope, except_for=None, on
     generic_members = set(generic_test_class.__dict__.keys()) - set(empty_class.__dict__.keys())
     generic_tests = [x for x in generic_members if x.startswith('test')]
 
+    # MPS backend support is disabled in `get_device_type_test_bases` while support is being ramped
```
Comment (mruberry) [inline torch/testing/_internal/common_device_type.py:615]: Nice comments

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -6258,7 +6258,7 @@ def make_bool_mask(shape):
 
         if mask_t.sum() == 0:
             def random_index(shape):
-                return tuple(map(lambda max_idx: random.randint(0, max_idx), shape))
+                return tuple(map(lambda max_idx: random.randrange(0, max_idx), shape))
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:6261]: Why the change from randint?

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -8702,6 +8704,7 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
                # TypeError: _copy_dispatcher() got an unexpected keyword argument 'memory_format'
                # (NumPy reference needs to be extended with memory_format)
                DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_numpy_ref'),
+               DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_numpy_ref_mps'),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:8707]: Seems like the list of skips should be converted into decorators like these?

**Review by mruberry — COMMENTED** [review]
This looks pretty good, @aslvrstn. I made a few inline comments for your review!

File: test/test_ops.py (hunk 1 above)
Comment (aslvrstn) [reply test/test_ops.py]: d'oh, i even found that and then didn't think to use it for this. Done in https://github.com/pytorch/pytorch/pull/87342/commits/a1dc7a862088ef5878375bd3122216f4c694dae9

File: test/test_ops.py (hunk 2 above)
Comment (aslvrstn) [reply test/test_ops.py:183]: If I do that, literally everything fails with
`TypeError: Cannot convert a MPS Tensor to float64 dtype as the MPS framework doesn't support float64. Please use float32 instead.`
so I'm assuming this is correct and it's a limitation of MPS? It's a real pain, too! The `broken_on_ref_inputs` are mostly failing on tests that test type promotion to float64

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (aslvrstn) [reply torch/testing/_internal/common_methods_invocations.py:6261]: I think this was a lurking bug! I can't find the CI run anymore that showed this failure, but it looks like this is trying to draw a valid index into a dimension of length max_idx, but randint draws inclusively to its upper bound, so for a length-5 dimension, it sometimes generates 5 as an index and gets an out of bounds issue.

Not totally sure how this wasn't triggering before

**Comment by aslvrstn** [comment]
I think I addressed everything. PTAL

File: test/test_ops.py (hunk 2 above)
Comment (mruberry) [reply test/test_ops.py:183]: That's understandable, yep. In that case I would also remove the setting of the default dtype within the test for consistency, and add a comment that, at the time the test was added, MPS didn't support double, which is why the test compares with NumPy in float32

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:6261]: That is odd but I see what you mean. 

cc @kshitij12345 @pmeier -- we should consider modifying `make_tensor` to return boolean tensors with a mix of true and false inputs, or providing a new `torch.testing.make_boolean_mask` function (name TBD) for this functionality

File: test/test_ops.py (hunk 2 above)
Comment (aslvrstn) [reply test/test_ops.py:183]: Done. This broke a few more tests which seem to mistakenly promote to float64, so I skipped those as well.

**Comment by mruberry** [comment]
This looks good to me, let's give @albanD a chance to review, too. Nice work, @aslvrstn!

**Review by albanD — COMMENTED** [review]
Sounds great! Thanks for taking the time to do this!
I think that beyond moving it to the mps-specific test file, this is all good for me! (sorry for asking to rename the skips now :( )

File: test/test_ops.py [hunk 6]
```diff
@@ -172,6 +173,24 @@ def test_numpy_ref(self, device, dtype, op):
...
+    def test_numpy_ref_mps(self, device, dtype, op):
```
[the remark this answers is not in the document]
Comment (aslvrstn) [reply test/test_ops.py]: Done in https://github.com/pytorch/pytorch/pull/87342/commits/473cca09d5cd4767e8295395e02b2eafd9d30fde

If you have ideas for unifying them again, though, I'm all ears. It's a kind of unfortunate amount of weird duplication, and I'm now realizing based on

```
# Get names of all the operators which have ref in their entry in OpInfo (testing infra)
#   except for elementwise unary operators (separately implemented in test/test_unary_ufuncs.py),
#   elementwise binary operators (separately implemented in test_binary_ufuncs.py),
#   reduction operations (separately impelemented in test_reductions.py),
#   and Spectral Functions (separately implemented for only 1D as of now, in test/test_spectral_ops.py)
```

that I'm about to get nerd-sniped into make _those_ run for MPS as well, and duplicating those tests into test_mps seems like a total nonstarter.

File: test/test_mps.py [hunk 7]
```diff
@@ -7790,10 +7809,56 @@ def req_grad(t):
...
+    # MPS still requires some fairly heavy special casing in the test framework.
+    # When MPS becomes more consistent, this can probably be merged with that test using
+    # `@dtypesIfMPS(torch.float32)`, but for now, the assertions themselves need to be loosened
+    @unittest.skipIf(TEST_WITH_ASAN, "Skipped under ASAN")
+    @onlyMPS
+    @suppress_warnings
+    # MPS only supports float32
+    @ops(_ref_test_ops, allowed_dtypes=(torch.float32,))
+    def test_numpy_ref_mps(self, device, dtype, op):
+        # Unlike above, this test compares in `float32` since at the time of this test's creation MPS does
```
Comment (albanD) [inline test/test_mps.py:7847]: nit this comment needs updating.

**Review by albanD — APPROVED** [review]
Thanks!
Let's merge this and build on top of it in follow up PRs!

**Comment by albanD** [comment]
(feel free to merge this using the pytorchbot merge command once typo is fixed!)

File: test/test_mps.py [hunk 8]
```diff
@@ -18,23 +18,42 @@
...
+)
+from torch.testing._internal.common_device_type import ops, instantiate_device_type_tests, onlyMPS
 from torch.testing._internal.common_nn import NNTestCase
 import numpy as np
 import torch
 import torch.utils._pytree as pytree
 
+
+# Copied from `test_ops.py` for the purposes of duplicating `test_numpy_ref`
+_ref_test_ops = tuple(
```
Comment (mruberry) [inline test/test_mps.py:47]: Does this really want to filter out elementwise unary, elementwise binary, reduction, and spectral operations? 

The reason those are excluded in test_ops.py is because they have more specific tests that run in their own test suites, like:

https://github.com/pytorch/pytorch/blob/d73d4aa7de953a3794593ac9e6d6b3a1ce514c3c/test/test_unary_ufuncs.py#L287

But if those test suites aren't run using the MPS device then it wouldn't be duplicative to include them in this ref test.

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:6268]: We can construct boolean tensors today but I don't think we guarantee that they have a mix of True and False, right? Like the tensor could be all True or all False?

File: test/test_mps.py (hunk 8 above)
Comment (aslvrstn) [reply test/test_mps.py:47]: I think the end goal should be that those tests also run on MPS, but that's definitely too large for this PR.

That's one reason I'm not thrilled having this live over in test_mps, tbh, see https://github.com/pytorch/pytorch/pull/87342#discussion_r1001112900

Either way, I'll tackle expanding the set of tests in a followup. All I wanted was to land a dang `where` test! :)

Comment (mruberry) [reply test/test_mps.py:47]: Sounds good!

File: torch/testing/_internal/common_methods_invocations.py (hunk 4 above)
Comment (pmeier) [reply torch/testing/_internal/common_methods_invocations.py:6260]: Fair point, but we sample with 50% probability for each

https://github.com/pytorch/pytorch/blob/620dbc43d8e3c836ad8d934987ee2f87fefbad7a/torch/testing/_creation.py#L129-L130

So unless you sample very small tensors, your chance to get an all `False` or `True` tensor back are vanishingly small. Like `0.5 ** (prod(shape) - 1)` small. If one has such a use case, i.e. a small mask that needs to have both `True` and `False`, they are probably better off hard-coding it.

I wouldn't act on this unless we get request for it and even than it is still up for debate IMO.
